azerothcore/azerothcore-wotlk
# refactor(Docker): remove prod container distinction

## Changes Proposed:
This PR proposes changes to:
- [x] Apps/DevOps

This PR removes the "prod" distinction from the containers, and now all containers have a self-contained build. Other changes that made it in:

- `docker-build.yaml` github action reworked to run the docker build in a more native way
- Refactor `Dockerfile` 
    - store the ccache in docker's cache
    - use a simpler build process with bare cmake rather than calling through the `./acore.sh` dashboard
    - Strive for smaller images overall
- `ac-dev-server` docker build stage split out into its' own Dockerfile
    - It didn't really fit in with the other Dockerfile
    - This allows for better logical separation with the dockerfiles
    - There's further improvement here to split out the current Dockerfile (so there's one for the worldserver, authserver, etc)
- Readme at `apps/docker/README.md` rewritten to better describe how to get started with Azerothcore in Docker
- SQL files removed from worldserver/authserver containers
    - the db-import is there, we can let it do its' job
    - This aligns with [factor 12 of a 12-factor-app](https://12factor.net/admin-processes)
        - Acore isn't a 12 factor app and a lot of effort will be needed to get it there (as if thats a goal), but it's nice to treat it like one
- `apps/docker/docker-cmd.sh` has had deprecation warnings added for the (now removed) prod services
- greatly minimize code in `docker-compose.yml`, removing uncommon or complex options
    - We shouldn't be running privileged containers
    - We shouldn't (really) be letting the container manage its' niceness
    - We shouldn't be using a script inside the container to restart the servers
    - profiles aren't necessary for this usecase
- Changes from #17405 

Goals of this PR:
- _simplicity_
- making the docker builds feel easier to work on and understand
- making the docker compose feel easier to work on
- remove extraneous settings and features used
    - For docker containers, _no frills is better as a default_
    - If someone wants or needs settings or configurations that were in the tree before this commit, they should add those in via an [override file](https://docs.docker.com/compose/multiple-compose-files/merge/) (which is in the gitignore so they should be ready to go) rather than making potentially complex changes to the default

this change was _supposed to be_ minimal; it was quite difficult on some things to decide what to keep and what to change.

## Issues Addressed:

- Supersedes #17405  

## SOURCE:

The changes have been validated through:

## Tests Performed:

This PR has been:


## How to Test the Changes:


### optional setup to test importing an existing database
1. Check out the `master` branch
1. build containers as normal (`./acore.sh docker build`)
2. start containers as normal (`./acore.sh docker start:app:d`)
3. Create an account as normal
4. Create a character, jump around, etc
5. check out this branch

### the actual test

1. run `docker compose up -d --build`
2. set your realmlist to localhost
3. create an account as normal (`docker compose attach ac-worldserver`, run `account create admin admin 3 -1`)
4. login to the game client
6. It should work without issues

## Known Issues and TODO List:


- `acore-docker` compatibility will need to be confirmed

- ~~I need to test this with importing a current docker installation; that is, if someone does a `git pull` and this pr had been merged, I need to confirm that this won't bust everything for them~~ Was able to import a build without issues
- ~~If this PR is merged, the Wiki will need some updates~~ updated https://github.com/azerothcore/wiki/pull/920 with matching information
- ~~the long, demoralizing task of testing and tweaking Github Actions~~ Seems to be working


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
